### ISSUE TITLE
throttle/quota on triggers

### ISSUE THREAD

**Issue by agnauck** [Azure/azure-webjobs-sdk#923]
Is there a way to throttle queue based function invokes?

For example if the queue triggered functions calls external APIs. Its very common that external API providers have quotas on how many calls you can make in a certain windows.

is there currently a solution for this in the webjobs SDK or extension packages?
if not I am looking for some guidance on how to implement it.

Thanks,
Alex

**Comment by lindydonna** [comment]
Related to https://github.com/Azure/azure-webjobs-sdk/issues/844

**Comment by brettsam** [comment]
There's not a direct mechanism to throttle this. The linked issue #844 will make it easier for you to control the invisibility timeout should a function fail.

But another option is to take the queue message, clone it, and re-add it to the queue with a visibility timeout. That means that it will not appear in the queue until that timeout expires. This lets you delay the retry on your own:

``` csharp
public static async Task QueueTrigger([QueueTrigger("test-input")] string message, [Queue("test-input")] CloudQueue queue)
{
    // if you need to throttle, insert a new message with a visibility timeout of 1 minute
    await queue.AddMessageAsync(new CloudQueueMessage(message), null, TimeSpan.FromMinutes(1), null, null);
}
```

**Comment by agnauck** [comment]
its not about delaying retries. It about setting quotas on queues.
eg: process max. 300 messages per minute on queueX

**Comment by lindydonna** [comment]
@agnauck We think the fix for #844 will resolve your issue. Basically, you'll be able to tweak the invisibility timeout to process queue messages more slowly. @mathewc or @brettsam can probably explain this better than I can. :)
